graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java

```diff
@@ -83,33 +73,49 @@ public StreamRouterEngine(@Assisted List<Stream> streams,
         this.streamProcessingTimeout = streamFaultManager.getStreamProcessingTimeout();
         this.fingerprint = new StreamListFingerprint(streams).getFingerprint();
 
-        for (final Stream stream : streams) {
-            for (final StreamRule streamRule : stream.getStreamRules()) {
+        this.rulesList = Lists.newArrayList();
+
+        final List<Rule> presenceRules = Lists.newArrayList();
+        final List<Rule> exactRules = Lists.newArrayList();
+        final List<Rule> greaterRules = Lists.newArrayList();
+        final List<Rule> smallerRules = Lists.newArrayList();
+        final List<Rule> regexRules = Lists.newArrayList();
+
+        for (Stream stream : streams) {
```

Comment:
Maybe instead of sorting the stream rules manually, it would make sense to introduce some kind of complexity rating for the different types of rules, and use a comparator to sort the list of rules accordingly.